tket2-hseries/src/extension/wasm.rs

```diff
@@ -0,0 +1,885 @@
+    match arg.into() {
+        TypeArg::Type { ty } => Some(ty.into()),
+        TypeArg::Variable { v } => Some(TypeRV::new_row_var_use(v.index(), v.bound_if_row_var()?)),
+        _ => None,
+    }
+}
+
+/// Fallibly convert a [TypeArg] to a [TypeRow].
+///
+/// This will fail if `arg` is of non-sequence kind (e.g. Type)
+/// or if the sequence contains row variables.
+/// TODO move this to `impl TypeArg` in `hugr-core`
+pub fn type_arg_into_type_row(arg: impl Into<TypeArg>) -> Option<TypeRow> {
+    match arg.into() {
+        TypeArg::Sequence { elems } => elems
+            .into_iter()
+            .map(type_arg_into_type)
+            .collect::<Option<Vec<_>>>()
+            .map(|x| x.into()),
+        _ => None,
+    }
+}
+
+/// Fallibly convert a [TypeArg] to a [TypeRowRV].
+///
+/// This will fail if `arg` is of non-sequence kind (e.g. Type).
+/// TODO move this to `impl TypeArg` in `hugr-core`
+pub fn type_arg_into_type_row_rv(arg: impl Into<TypeArg>) -> Option<TypeRowRV> {
+    match arg.into() {
+        TypeArg::Sequence { elems } => elems
+            .into_iter()
+            .map(type_arg_into_type_rv)
+            .collect::<Option<Vec<_>>>()
+            .map(|x| x.into()),
+        TypeArg::Variable { v } => {
+            Some(vec![TypeRV::new_row_var_use(v.index(), v.bound_if_row_var()?)].into())
+        }
+        _ => None,
+    }
+}
+
+/// TODO move this to `impl TypeArg` in `hugr-core`
+fn type_row_into_type_arg(row: TypeRow) -> TypeArg {
+    TypeArg::Sequence {
+        elems: row.into_owned().into_iter().map_into().collect(),
+    }
+}
+
+/// TODO move this to `impl TypeArg` in `hugr-core`
+fn type_row_rv_into_type_arg(row: TypeRowRV) -> TypeArg {
+    TypeArg::Sequence {
+        elems: row.into_owned().into_iter().map_into().collect(),
+    }
+}
+
+impl MakeOpDef for WasmOpDef {
+    fn init_signature(&self, extension_ref: &Weak<Extension>) -> SignatureFunc {
+        let context_type = WasmType::Context.get_type(extension_ref);
+        let module_type = WasmType::Module.get_type(extension_ref);
+        match self {
+            Self::get_context => Signature::new(
+                usize_t(),
+                Type::from(WasmOp::get_context_return_type(extension_ref)),
+            )
+            .into(),
+            Self::dispose_context => Signature::new(context_type, type_row![]).into(),
+            Self::lookup => {
+                let inputs = TypeRV::new_row_var_use(1, TypeBound::Copyable);
+                let outputs = TypeRV::new_row_var_use(2, TypeBound::Copyable);
+
+                let func_type = WasmType::func_custom_type(inputs, outputs, extension_ref).into();
+                PolyFuncTypeRV::new(
+                    [
+                        NAME_PARAM.to_owned(),
+                        INPUTS_PARAM.to_owned(),
+                        OUTPUTS_PARAM.to_owned(),
+                    ],
+                    Signature::new(vec![module_type], vec![func_type]),
```

Comment:
I'm a noob when it comes to define (parametrised) type signatures, so take this with a grain of salt. I notice that `Self::lookup` and `Self::call` use different ways to define the signatures returned, even though superficially they seem quite similar to me (`Signature::new` vs `FuncValueType::new`, `Type` vs `WasmType` etc)...

It would make it more readable if both code blocks were as similar as possible. If you can't use the same function calls & types for whatever reason, then maybe just put the respective function signatures in a comment -- it took me awhile to parse what they meant and had to refer several times back to the module docs (which were supremely useful btw).